### PR TITLE
[4.0] Frontend Template Cassopeia - Add submenu dropdown 

### PR DESCRIPTION
**Pull request by coolcat-creations**
Pull Request for Issue  #29025 .

### Summary of Changes
Adding Submenu css to the menu

### Testing Instructions
not for final testing yet
collaboration is welcome

open issues:
- mobile submenu
- maximum levels?
- javascript needed?

### Expected result
Multilevel Dropdown in cassopeia possible

grafik

### Actual result
no submenu possible at all

### CODE REVIEW

**Comment by richard67** [comment]
@coolcat-creations Currently it seems to work only with 2 levels. If having 3 menu item levels, the 2nd level dropdown properly opens if having selected the first level, but the dropdown for the 3rd level doesn't open when selecting the 2nd level menu item in the 2nd level drop down.

**Comment by coolcat-creations** [comment]
I am not very satisfied about it at all and asked someone from the GER Community to help, I hope a better PR will follow. I would know how to solve it for client projects but here it has to work generally for everything possible and OC withouth jQuery - I don't know how to solve it in the best way :-)

**Comment by richard67** [comment]
I wish I could help, but I'm not expert in this stuff.

**Comment by ChristineWk** [comment]
@coolcat-creations 
I'm afraid - can't also help further ....
Tried something and got following result (screenshot) maybe belongs to scss - NPM needed?

@richard67 
It was possible, when selecting the 3rd sub-menue to select the 2nd sub-menue.

OT or not - bug or feature :-) with following steps:

a) "Parent Item" > e.g. Blog
b) "Menue Item Type" > ...... etc. (I took single article only)
c) "The Parent Item" changed back to: "Menu Item Root"
d) so, to start again with a)

submenues

**Comment by richard67** [comment]
@ChristineWk For this PR here you need to run npm, that's right.

**Comment by dgrammatiko** [comment]
> javascript needed?

JS is needed at least for accessibility (would be nice that the code would be BS js free as well). FWIW whenever a dropdown opens you need to set some aria-expanded, take care of the the tabstate and focus on the first `a` tag. As is, this completely breaks a11y...

**Comment by brianteeman** [comment]
luckily bootstrap 4 deals with all of that for me ;)

**Comment by astridx** [comment]
I also believe that we cannot deliver the Cassiopeia menu as it is at the moment.
Does anyone know where and why it was decided that Boostrap is not used in Cassiopeia? And why are we still loading Boostrap in the head of Cassiopeia?

**Comment by C-Lodder** [comment]
@astridx What parts of Cassiopeia aren't using BS apart from the core grid?

**Comment by astridx** [comment]
> @C-Lodder What parts of Cassiopeia aren't using BS apart from the core grid?

I ask because I understood [Brian's comment](https://github.com/joomla/joomla-cms/pull/29413#issuecomment-639102090) that Cassiopeia shouldn't use Boostrap. Did I misunderstand that?

**Comment by C-Lodder** [comment]
@astridx He was saying that BS takes care of the accessibility for dropdowns.

**Comment by ckubs** [comment]
Making already something nice out of Cassiopeia( on beta 2) decided to stick with it for my new site. Bumped on this thread as the only issue I have is the lack of dropdowns for the main menu.

I managed to enable dropdowns using [this snippet](https://codepen.io/piscu/pen/EbWQda/), on click. Unfortunatelly I don't have the necesarry jQuery knowledge to make a **span** inside **.nav-item.deeper** clickable to open the dropdown. Instead of the actual link wich may be a parent category that should open it's URL.

If someone knows wether Cassiopeia wil have proper menu dropdowns officialy to wait a littlew or is willing to give a little help, I'm gratefull. :)

**Comment by astridx** [comment]
Please see https://github.com/joomla/joomla-cms/pull/29929

**Comment by david0296** [comment]
I have tested this item :white_check_mark: successfully on 0116f3800b2cc5e050e7d957562bb4aefb564692<br /><br />After applying the pr, I was able to create multi-level dropdown-menus as shown in the screenshot<hr />
joomla
<sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29413">issues.joomla.org/tracker/joomla-cms/29413</a>.</sub>

File: templates/cassiopeia/scss/blocks/_header.scss [hunk 1]
```diff
@@ -75,16 +75,82 @@
   .mod-menu li {
     position: relative;
     display: inline-block;
-    margin-right: 20px;
+    margin-right: 1.5rem;
     text-transform: uppercase;
     letter-spacing: 1.75px;
+    cursor: pointer;
+
+
```
Comment (brianteeman) [inline templates/cassiopeia/scss/blocks/_header.scss:82]:
```suggestion
```

File: templates/cassiopeia/scss/blocks/_header.scss [hunk 2]
```diff
@@ -75,16 +75,82 @@
...
+        }
+    }
+
+    &.dropdown > .dropdown-menu  {
+      display: none;
+      opacity: 0;
+      transition: opacity 0.5s ease;
+      padding: 0.5rem;
+
+
```
Comment (brianteeman) [inline templates/cassiopeia/scss/blocks/_header.scss:102]:
```suggestion
```

File: templates/cassiopeia/scss/blocks/_header.scss [hunk 3]
```diff
@@ -75,16 +75,82 @@
...
+    &.dropdown:hover > .dropdown-menu {
+        display: block;
+        opacity: 1;
+        margin-top: 0;
+        border: none;
+
+        @include media-breakpoint-up(md) {
+          -webkit-box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.6);
+          -moz-box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.6);
+          box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.6);
```
Comment (brianteeman) [inline templates/cassiopeia/scss/blocks/_header.scss:92]:
```suggestion
        @include media-breakpoint-up(md) {
          -webkit-box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.6);
          -moz-box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.6);
          box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.6);
```

**Comment by infograf768** [comment]
RTL ?

**Comment by coolcat-creations** [comment]
I don't know if I made it clear in the initial post. This PR is far from being good. It would be great if someone could help to create a good menu with a dropdown or even better a megamenu. In Glip we have a public chat about cassiopeia anyone can join to discuss how to improve it.

**Comment by chmst** [comment]
Work in progress for cassiopeia.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29413">issues.joomla.org/tracker/joomla-cms/29413</a>.</sub>

**Comment by brianteeman** [comment]
If you change the pr to draft  status to indicate it is still a work in progress it would be helpful

**Comment by Magnytu2** [comment]
Hello everyone, once again I'm playing with J4 and the menus. But there is no change for the submenu. Do you have a close solution?
Positions_des_modules-3

**Comment by infograf768** [comment]
@Magnytu2 
Cirylle, please read "**Draft**" near the title. It means this PR is not ready. Be patient.

**Comment by Magnytu2** [comment]
> @ Magnytu2
> Cirylle, veuillez lire " **Brouillon** " à côté du titre. Cela signifie que ce PR n'est pas prêt. Sois patient.

Sorry, here it is no longer the patient with Joomla! 4. It gets depressing. I want to help the community, to show the progress of Joomla, but each time I am disappointed. It becomes difficult to support Joomla.

**Comment by richard67** [comment]
As said, someone is working on it and we will see results soon. Stay tuned.

**Comment by Magnytu2** [comment]
> Comme dit, quelqu'un y travaille et nous verrons bientôt des résultats. Restez à l'écoute.

Thank you, I am. My Joomanji.fr site is just there for that. To show, help, create with Joomla 4. Except I don't know what to show and how to encourage Joomla?

**Comment by web-eau-net** [comment]
This PR have been created by coolcat-creations the 3rd June. 
Today is 16th Aug.
Please be patient @Magnytu2 be very patient.... :-1:

**Comment by richard67** [comment]
Well you guys write that you want to know how to help. I will soon provide a link to our ongoing work where everybody can discuss and contribute. But complaining again doesn't help, I think. If there has been something said before, it is not useful to repeat and repeat it again.

**Comment by web-eau-net** [comment]
3rd Jun - 16th Aug,...

**Comment by Magnytu2** [comment]
I'm not a developer, but I'll be happy to test your changes. Not with Joomla's PRs and the like. It's completely inaccessible to users like me. But if you have a template in .zip format that I can install, and set up several sites to make simulations, it will be easy.

**Comment by astridx** [comment]
> I'm not a developer, but I'll be happy to test your changes. Not with Joomla's PRs and the like. It's completely inaccessible to users like me. But if you have a template in .zip format that I can install, and set up several sites to make simulations, it will be easy.

@Magnytu2  I agree with you. Testing a PR is complicated. But what exactly do you mean with "Not with Joomla's PRs and the like".  Do you know that you can download a zipped version of Joomla -in PR available- for testing? And if no JS or SCSS File is changed you mostly can use https://github.com/joomla-extensions/patchtester.

4 0 Frontend Template Cassopeia Add submenu dropdown by coolcat creations · Pull Request 29413 · joomla joomla cms(1)

**Comment by astridx** [comment]
> This PR have been created by coolcat-creations the 3rd June.
> Today is 16th Aug.
> Please be patient @Magnytu2 be very patient.... -1

@web-eau-net 
A long period of time, that's true. But who do you think is responsible for building the drop down?

By the way, I submitted a working solution. It is linked in this PR. My PR has not been tested. I received two responses. In my opinion that is little. As I see it, both respnses are not relate to the code, rather it was more about preferences (Boostrap, Query or not, no further dependency without presenting the solution who alternatively programmed ...)

**Comment by Magnytu2** [comment]
@astridx No I am sorry. Github is an occult place, completely inaccessible and incomprehensible. Even for me who defines myself as a handyman, I dare not touch or say anything. For fear of doing stupid things or offending someone. I'm only testing what I can figure out by installing a Joomla or extension. And I'm just screaming at my sadness to see Joomla bogged down, losing users, and failing to communicate.

**Comment by bato3** [comment]
@coolcat-creations it also require changes in `/media/mod_menu/js/menu.js` to add keyboard support for dropdown open

**Comment by ceford** [comment]
For my own J4 needs I made a clone of Cassiopeia and used Smartmenus (https://www.smartmenus.org/) which requires jQuery. I needed to revise the header and banner area layout to - but that is an aside. I only wanted one drop-down level, although it could do more. I am quite satisfied with the result - responsive and accessible. I am not suggesting that Joomla should use it, just that there are other solutions out there.

screen shot 2020-11-18 at 12 11 29<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29413">issues.joomla.org/tracker/joomla-cms/29413</a>.</sub>

**Comment by richard67** [comment]
@ceford The [frontend-template-working-group](https://volunteers.joomla.org/teams/frontend-template-working-group) is on the way to finish their works on [https://github.com/joomla/cassiopeia](https://github.com/joomla/cassiopeia), which will result in a pull request for the 4.0-dev branch here in the CMS repository. The work includesd an accessible drop-down menu based on metismenujs, the vanilla JS version of metismenu. Metismenujs is already shipped with the CMS for the backend, and so it doesn't introduce a new dependency on another 3rd party product.

**Comment by richard67** [comment]
P.S.: We don't want to introduce another dependency on jQuery and another external dependency to a 3rd party product.

**Comment by ceford** [comment]
I just remembered that I sought an alternative menu solution because the then Cassiopeia menu was not suitable for my needs. That must have been early this year. I look forward to testing the metsmenu based menu in due course.

**Comment by chmst** [comment]
@coolcat-creations is this still work in progress for you?

**Comment by richard67** [comment]
The [Frontend Template Working Group (JEDT - FT)](https://volunteers.joomla.org/teams/frontend-template-working-group) has finished their works on the [https://github.com/joomla/cassiopeia](https://github.com/joomla/cassiopeia) repository and just created PR #31520 , which includes also an accessible dropdown menu for the frontend.

@coolcat-creations and others: I'd be happy if you could test our PR.

**Comment by Quy** [comment]
@richard67 Close?

**Comment by richard67** [comment]
@Quy In principle yes, because now after the Cassiopeia enhancements with the dropdown menu has been merged, this PR is obsolete. On the other hand I prefer to leave that to authors.

**Pull request closed without merging**